modelci/finetuner/trainer.py

```diff
@@ -0,0 +1,232 @@
+
+    val_loss, val_corrects = 0., 0
+    pbar = tqdm(total=epochs * (len(trainer.train_data_loader) + len(trainer.test_data_loader)))
+    for epoch in range(epochs):
+        trainer.net.train()
+        num_samples = 0
+        train_loss, train_corrects = 0., 0
+        running_loss, running_corrects = 0., 0
+
+        for batch_num, samples in enumerate(trainer.train_data_loader):
+            trainer.optimizer.zero_grad()
+            num_samples += trainer.get_sample_size(samples)
+            loss, corrects = trainer.train_one_batch(samples)
+            running_loss += loss
+            train_loss += loss
+            running_corrects += corrects
+            train_corrects += corrects
+
+            pbar.update(1)
+
+            if log_batch_num is not None and batch_num % log_batch_num == 0:
+                pbar.set_description(
+                    f'[epoch {epoch:d}] [Training step {batch_num:d}] '
+                    f'train loss {running_loss / num_samples:2.5g} | acc {running_corrects / num_samples:2.5g} '
+                    f'|| test loss {val_loss:2.5g} | acc {val_corrects:2.5g}'
+                )
+                num_samples = 0
+                running_loss = 0.
+                running_corrects = 0
+
+        train_loss /= trainer.train_data_size
+        train_corrects /= trainer.train_data_size
+
+        pbar.set_description(
+            f'[epoch {epoch:d}] [Evaluating] '
+            f'train loss {train_loss:2.5g} | acc {train_corrects:2.5g} '
+            f'|| test loss --- | acc ---'
+        )
+
+        trainer.net.eval()
+        val_loss, val_corrects = 0., 0
+
+        for samples in trainer.test_data_loader:
+            loss, corrects = trainer.evaluate_one_batch(samples)
+            val_loss += loss
+            val_corrects += corrects
+            pbar.update(1)
+
+        val_loss /= trainer.test_data_size
+        val_corrects /= trainer.test_data_size
+
+        pbar.set_description(
+            f'[epoch {epoch:d}] [Saving Model] '
+            f'train loss {train_loss:2.5g} | acc {train_corrects:2.5g} '
+            f'|| test loss {val_loss:2.5g} | acc {val_corrects:2.5g}'
+        )
+
+        # process statistics
+        train_losses[epoch], train_accs[epoch] = train_loss, train_corrects
+        val_losses[epoch], val_accs[epoch] = val_loss, val_corrects
+
+        # save model
+        if val_loss < best_val_loss:
+            pbar.set_description(
+                f'[epoch {epoch:d}] [Done] '
+                f'train loss {train_loss:2.5g} | acc {train_corrects:2.5g} '
+                f'|| test loss {val_loss:2.5g} | acc {val_corrects:2.5g}'
+            )
+            best_val_loss = val_loss
+            torch.save(net.state_dict(), save_path)
+
+    stat = {'train_loss': train_losses, 'train_acc': train_accs, 'test_loss': val_losses, 'test_acc': val_accs}
+
+    return {'stat': stat, 'info': {'save_name': save_path}}
+
+
+class CIFAR10Trainer(Trainer):
```

Comment:
We do not need this class.

Dataset name should be parameter can read by the trainer.

You can keep your experiments. I will help to design this.

Reply:
Add a todo and merge.

I am going to refactor this directory next week.

Reply:
Added a TODO